README.txt
= StringValidator =

文字列の形式/値を検証する。

== 作者 ==

とみたまさひろ <tommy@tmtm.org>

== ライセンス ==

Ruby ライセンス http://www.ruby-lang.org/ja/LICENSE.txt と同等。

== 機能 ==

 * 文字列が規則に従うかどうかの検査。
 * 規則は Ruby のリテラルで記述する。

== ダウンロード ==

 * http://tmtm.org/downloads/ruby/stringvalidator/

== インストール ==

{{{
$ make
$ make test
# make install
}}}

== 使用例 ==

str が "abc"？
{{{
StringValidator.valid?("abc", str)
}}}

str が正規表現に一致？
{{{
StringValidator.valid?(/hoge/, str)
}}}

str が 123 の数値？
{{{
StringValidator.valid?(123, str)
}}}

str が 1〜255 の整数？
{{{
StringValidator.valid?(1..255, str)
}}}

str が 1〜255 の数値？
{{{
StringValidator.valid?(1.0..255, str)
}}}

str が "abc" または "def"？
{{{
StringValidator.valid?(["abc", "def"], str)
}}}

str の長さが 3〜10？
{{{
StringValidator.valid?({:length=>3..10}, str)
}}}

str の長さが 10文字以上で正規表現に一致？
{{{
StringValidator.valid?({:minlength=>10, rule=>/abc/}, str)
}}}

ブロックを呼び出した結果が真？
{{{
p = Proc.new{|a| a == "xyz"}
StringValidator.valid?(p, "xyz")
}}}

複数のルールを定義して利用
{{{
rules = {
  :port => 1..65535,
  :domain => /\A[a-z0-9-]+(\.[a-z0-9-]+)+\z/i
}
v = StringValidator.new(rules)
v.valid?(:port, "8080")
v.valid?(:domain, "tmtm.org")
}}}

どのルールに適合したか？
{{{
rules = {
  :port => 1..65535,
  :domain => /\A[a-z0-9]+(\.[a-z0-9]+)+\z/i
}
v = StringValidator.new(rules)
v.validated_rule("8080")     # => :port
v.validated_rule("tmtm.org") # => :domain
v.validated_rule("xyz")      # => nil
}}}

=== メソッド ===

==== self.validate(rule, str) ====

str が rule で評価し、適合すれば評価結果を返す。適合しなければ例外を発生させる。

rule は次のように評価される。

 Integer::
  整数かどうか。Integer(str) を返す。

  例外:
  * StringValidator::Error::NotInteger

 Float::
  実数かどうか。Float(str) を返す。

  例外:
  * StringValidator::Error::NotFloat

 Rangeオブジェクト::
  rule.first と rule.last が整数であれば、rule.include?(str.to_i)。str.to_i を返す。

  例外:
   * StringValidator::Error::OutOfRange (範囲外)
   * StringValidator::Error::NotInteger (str が整数でない)

  rule.first が実数であれば、rule.include?(str.to_f)。str.to_f を返す。

  例外:
   * StringValidator::Error::OutOfRange (範囲外)
   * StringValidator::Error::NotFloat (str が実数でない)

  それ以外の場合、rule.include?(str)。str を返す。

  例外:
   * StringValidator::Error::OutOfRange (範囲外)

 Regexpオブジェクト::
  rule =~ str が真。str を返す。

  例外:
   * StringValidator::Error::RegexpMismatch (非適合)

 Procオブジェクト::
  rule.call(str) が真。rule.call(str) の戻り値を返す。

  例外:
   * StringValidator::Error::InvalidValue (結果が偽)

 Arrayオブジェクト::
  rule の各要素について valid? で評価し、ひとつでも真であれば真。真になった rule の評価結果を返す。

  例外:
   * StringValidator::Error::InvalidValue (結果が偽)

 Hashオブジェクト::
  Hash の各要素について以下を評価し、すべてが真であれば真

  :any => [ ... ]
    配列の各要素について validate で評価し、ひとつでも真であれば真。Array と同じ。
  :all => [ ... ]
    配列の各要素について validate で評価し、すべてが真であれば真。最後の評価結果を返す。
  :rule => obj
    obj を validate で評価。評価結果を返す。
  :length => obj
    str.length について obj で評価。str を返す。
  :maxlength => n
    str のバイト数が n 以下であれば真。str を返す。
  :minlength => n
    str のバイト数が n 以上であれば真。str を返す。
  :maxcharlength => n
    str の文字数が n 以下であれば真。str を返す。文字の扱いは $KCODE に依存する。
  :mincharlength => n
    str の文字数が n 以上であれば真。str を返す。文字の扱いは $KCODE に依存する。

  例外:
   * StringValidator::Error::InvalidValue (:ary に非適合)
   * StringValidator::Error::InvalidLength (:length に非適合)
   * StringValidator::Error::TooLong (:maxlength, :maxcharlength に非適合)
   * StringValidator::Error::TooShort (:minlength, :mincharlength に非適合)

 その他クラス::
  rule.new(str) が成功すれば真。rule.new(str) の結果を返す。

  例外:
   * StringValidator::Error::InvalidValue (非適合)

 上記以外::
  rule.to_s == str を評価。rule を返す。

  例外:
   * StringValidator::Error::InvalidValue (非適合)

==== self.valid?(rule, str) ====

self.validate(rule, str) と同じ評価を行ない、結果を true / false で返す。

==== self.new(rule_list) ====

rule_list のルール群を持つ StringValidator オブジェクトを生成する。

rule_list は Hash で以下の形式。
{{{
  {
    name => rule, ...
  }
}}}

rule は self.validate() の第１引数と同じ形式で規則を指定する。

==== validate(rule_name, str) ====

rule_list[rule_name] の規則で str を評価し、評価結果を返す。
適合しない場合は StringValidator::Error::* の例外が発生する。

==== valid?(rule_name, str) ====

StringValidator#validator(rule_name, str) と同じだが、結果を true / false で返す。

==== validated_rule(str) ====

rule_list の各規則で str を評価し、一致した結果の name を返す。
どの規則にも一致しない場合は nil を返す。

rule_list の規則は name の順に評価される。


HISTORY
== 0.4 2009-09-08 ==
 * Proc の評価結果が false の場合でも不正とみなさないように変更
 * StringValidator#validate で Symbol オブジェクトを規則として評価するようにした

== 0.3.3 2008-01-19 ==
 * :all の評価結果を、最初の要素の評価結果に変更
 * Hash の評価結果を、:rule, :any, :all の評価結果に変更

== 0.3.2 2007-12-09 ==
 * :all の評価結果を、要素すべての評価結果の配列に変更
 * rcov が 100% になるよう調整
 * Error.new の引数省略可

== 0.3.1 2007-12-05 ==
 * rule が Class クラスの実装
 * :charlength ルール追加
 * rdoc 化
 * Hash オブジェクトの時の validate の結果は Hash を返すようにした

== 0.3 2007-09-26 ==
 * validate でオブジェクトを返すようにした
 * :maxcharlength, :mincharlength 追加

== 0.2 2007-01-24 ==
 * :length 追加
 * :rule が働かなかったバグを修正
 * valid?() 追加
 * validate() で例外を発生するように変更
 * Float 追加
 * Range の最初と最後が Integer であれば整数として処理
 * Array 追加

== 0.1 2007-01-21 ==


spec/stringvalidator_spec.rb
# -*- encoding: utf-8 -*-
# Copyright (C) 2009 TOMITA Masahiro
# mailto:tommy@tmtm.org

require "pathname"
require File.expand_path "../lib/stringvalidator", File.dirname(__FILE__)

describe StringValidator do
  before do
    unless "".respond_to? :encoding
      @kcode = $KCODE
      $KCODE = "U"
    end
  end

  after do
    $KCODE = @kcode if @kcode
  end

  describe '.validate' do
    it 'Integer' do
      StringValidator.validate(Integer, "123").should == 123
      StringValidator.validate(Integer, "0").should == 0
      StringValidator.validate(Integer, "-213").should == -213
      lambda{StringValidator.validate(Integer, "1.2")}.should raise_error StringValidator::Error::NotInteger, 'not integer'
      lambda{StringValidator.validate(Integer, "a")}.should raise_error StringValidator::Error::NotInteger, 'not integer'
    end

    it 'Float' do
      StringValidator.validate(Float, "1.23").should == 1.23
      StringValidator.validate(Float, "123").should == 123
      StringValidator.validate(Float, "0").should == 0
      StringValidator.validate(Float, "-213").should == -213
      lambda{StringValidator.validate(Float, "a")}.should raise_error StringValidator::Error::NotFloat, 'not float'
    end

    it 'Integer const' do
      r = 987
      StringValidator.validate(r, "987").should == 987
      lambda{StringValidator.validate(r, "986")}.should raise_error StringValidator::Error::InvalidValue, 'invalid value'
    end

    it 'Range of Integer' do
      r = 3..45
      StringValidator.validate(r, "3").should == 3
      StringValidator.validate(r, "45").should == 45
      StringValidator.validate(r, "10").should == 10
      lambda{StringValidator.validate(r, "8.7")}.should raise_error StringValidator::Error::NotInteger, 'not integer'
      lambda{StringValidator.validate(r, "2.9")}.should raise_error StringValidator::Error::NotInteger, 'not integer'
      lambda{StringValidator.validate(r, "46")}.should raise_error StringValidator::Error::OutOfRange, 'out of range'
      lambda{StringValidator.validate(r, "hoge")}.should raise_error StringValidator::Error::NotInteger, 'not integer'
    end

    it 'Range of Numeric' do
      r = 3.0 .. 45
      StringValidator.validate(r, "3").should == 3
      StringValidator.validate(r, "45").should == 45
      StringValidator.validate(r, "10").should == 10
      StringValidator.validate(r, "8.7").should == 8.7
      lambda{StringValidator.validate(r, "2.9")}.should raise_error StringValidator::Error::OutOfRange, 'out of range'
      lambda{StringValidator.validate(r, "46")}.should raise_error StringValidator::Error::OutOfRange, 'out of range'
      lambda{StringValidator.validate(r, "hoge")}.should raise_error StringValidator::Error::NotFloat, 'not float'
    end

    it 'Range of String' do
      r = "abc" .. "xyz"
      StringValidator.validate(r, "abc").should == "abc"
      StringValidator.validate(r, "lmn").should == "lmn"
      StringValidator.validate(r, "xyz").should == "xyz"
      lambda{StringValidator.validate(r, "abb")}.should raise_error StringValidator::Error::OutOfRange, 'out of range'
      lambda{StringValidator.validate(r, "xyzz")}.should raise_error StringValidator::Error::OutOfRange, 'out of range'
    end

    it 'String' do
      r = "hogehoge"
      StringValidator.validate(r, "hogehoge").should == "hogehoge"
      lambda{StringValidator.validate(r, "123")}.should raise_error StringValidator::Error::InvalidValue, 'invalid value'
    end

    it 'Regexp' do
      r = /^[abc]$/
      StringValidator.validate(r, "a").should == "a"
      StringValidator.validate(r, "b").should == "b"
      StringValidator.validate(r, "c").should == "c"
      StringValidator.valid?(r, "abc").should == false
      StringValidator.valid?(r, "A").should == false
      lambda{StringValidator.validate(r, "abc")}.should raise_error StringValidator::Error::RegexpMismatch, 'regexp mismatch'
      lambda{StringValidator.validate(r, "A")}.should raise_error StringValidator::Error::RegexpMismatch, 'regexp mismatch'
    end

    it 'Array' do
      a = [123, "abc", /xyz/i]
      StringValidator.validate(a, "123").should == 123
      StringValidator.validate(a, "abc").should == "abc"
      StringValidator.validate(a, "xxxXyZzzz").should == "xxxXyZzzz"
      lambda{StringValidator.validate(a, "789")}.should raise_error StringValidator::Error::InvalidValue, 'invalid value'
    end

    it ':any' do
      any = {:any => [1, 2, 4, 8]}
      StringValidator.validate(any, "1").should == 1
      StringValidator.validate(any, "2").should == 2
      StringValidator.validate(any, "4").should == 4
      StringValidator.validate(any, "8").should == 8
      lambda{StringValidator.validate(any, "5")}.should raise_error StringValidator::Error::InvalidValue, 'invalid value'
      lambda{StringValidator.validate(any, "0")}.should raise_error StringValidator::Error::InvalidValue, 'invalid value'
    end

    it ':all' do
      all = {:all => [Integer, 5..10]}
      StringValidator.validate(all, "5").should == 5
      StringValidator.validate(all, "7").should == 7
      StringValidator.validate(all, "10").should == 10
      lambda{StringValidator.validate(all, "8.5")}.should raise_error StringValidator::Error::NotInteger, 'not integer'
      lambda{StringValidator.validate(all, "abc")}.should raise_error StringValidator::Error::NotInteger, 'not integer'
    end

    it ':rule' do
      hash = {:rule => /abc/}
      StringValidator.validate(hash, "012abc345").should == "012abc345"
      lambda{StringValidator.validate(hash, "12345")}.should raise_error StringValidator::Error::RegexpMismatch, 'regexp mismatch'
    end

    it ':length' do
      hash = {:length => 3}
      StringValidator.validate(hash, "123").should == "123"
      lambda{StringValidator.validate(hash, "1234")}.should raise_error StringValidator::Error::InvalidLength, 'invalid length'
    end

    it ':length as Range' do
      hash = {:length => 3..10}
      StringValidator.validate(hash, "123").should == "123"
      StringValidator.validate(hash, "1234567890").should == "1234567890"
      StringValidator.validate(hash, "12345").should == "12345"
      lambda{StringValidator.validate(hash, "12")}.should raise_error StringValidator::Error::InvalidLength, 'invalid length'
      lambda{StringValidator.validate(hash, "12345678901")}.should raise_error StringValidator::Error::InvalidLength, 'invalid length'
    end

    it ':minlength' do
      hash = {:minlength => 3}
      StringValidator.validate(hash, "1234").should == "1234"
      StringValidator.validate(hash, "123").should == "123"
      lambda{StringValidator.validate(hash, "12")}.should raise_error StringValidator::Error::TooShort, 'too short'
    end

    it ':maxlength' do
      hash = {:maxlength => 10}
      StringValidator.validate(hash, "123456789").should == "123456789"
      StringValidator.validate(hash, "1234567890").should == "1234567890"
      lambda{StringValidator.validate(hash, "12345678901")}.should raise_error StringValidator::Error::TooLong, 'too long'
    end

    it ':charlength' do
      hash = {:charlength => 3}
      StringValidator.validate(hash, "１２３").should == "１２３"
      lambda{StringValidator.validate(hash, "１２")}.should raise_error StringValidator::Error::InvalidLength, 'invalid length'
      lambda{StringValidator.validate(hash, "１２３４")}.should raise_error StringValidator::Error::InvalidLength, 'invalid length'
    end

    it ':charlength as Range' do
      hash = {:charlength => 3..10}
      StringValidator.validate(hash, "１２３").should == "１２３"
      StringValidator.validate(hash, "１２３４５６７８９０").should == "１２３４５６７８９０"
      StringValidator.validate(hash, "１２３４５").should == "１２３４５"
      lambda{StringValidator.validate(hash, "１２")}.should raise_error StringValidator::Error::InvalidLength, 'invalid length'
      lambda{StringValidator.validate(hash, "１２３４５６７８９０１")}.should raise_error StringValidator::Error::InvalidLength, 'invalid length'
    end

    it ':mincharlength' do
      hash = {:mincharlength => 3}
      StringValidator.validate(hash, "１２３４").should == "１２３４"
      StringValidator.validate(hash, "１２３").should == "１２３"
      lambda{StringValidator.validate(hash, "１２")}.should raise_error StringValidator::Error::TooShort, 'too short'
    end

    it ':maxcharlength' do
      hash = {:maxcharlength => 10}
      StringValidator.validate(hash, "１２３４５６７８９").should == "１２３４５６７８９"
      StringValidator.validate(hash, "１２３４５６７８９０").should == "１２３４５６７８９０"
      lambda{StringValidator.validate(hash, "１２３４５６７８９０１")}.should raise_error StringValidator::Error::TooLong, 'too long'
    end

    it 'multiple rule' do
      hash = {
        :maxlength => 5,
        :rule => /\A\d+\z/,
      }
      str = "12345"
      StringValidator.validate(hash, str).should == str
    end

    it ':rule, :any, :all' do
      hash = {
        :rule => proc{123},
        :any => ["12345", "abcde"],
        :all => [Integer, String],
      }
      str = "12345"
      StringValidator.validate(hash, str).should == 123
    end

    it 'invalid hash' do
      hash = {:hoge => nil}
      begin
        lambda{StringValidator.validate(hash, "hoge")}.should raise_error ArgumentError, 'Invalid key: hoge'
      end
    end

    it 'Proc' do
      p = Proc.new{|a| a == "xyz" && 123}
      StringValidator.validate(p, "xyz").should == 123
      StringValidator.validate(p, "abc").should == false
      p2 = Proc.new{raise "hoge"}
      lambda{StringValidator.validate(p2, "abc")}.should raise_error StringValidator::Error::InvalidValue, 'invalid value'
    end

    it 'Pathname' do
      r = Pathname
      StringValidator.validate(r, "abcdefg").should be_kind_of Pathname
      lambda{StringValidator.validate(r, "abcd\0efg")}.should raise_error StringValidator::Error::InvalidValue, 'invalid value'
    end
  end

  it '#validate' do
    rule = {
      :hoge => Integer,
      :fuga => "abc",
    }
    s = StringValidator.new(rule)
    s.validate(:hoge, "123").should == 123
    s.validate(:fuga, "abc").should == "abc"
    s.validate(Integer, "123").should == 123
    s.validate([:hoge, :fuga], "abc").should == "abc"
    lambda{s.validate(:hoge, "abc")}.should raise_error StringValidator::Error::NotInteger, 'not integer'
    lambda{s.validate(:fuga, "123")}.should raise_error StringValidator::Error::InvalidValue, 'invalid value'
    lambda{s.validate(:hage, "123")}.should raise_error ArgumentError, 'No such rule: hage'
  end

  it '#valid?' do
    rule = {
      :hoge => Integer,
      :fuga => "abc",
    }
    s = StringValidator.new(rule)
    s.valid?(:hoge, "123").should == true
    s.valid?(:fuga, "abc").should == true
    s.valid?(:hoge, "abc").should == false
    s.valid?(:fuga, "123").should == false
  end

  it '#validated_rule' do
    rule = {
      :hoge => Integer,
      :fuga => "abc",
    }
    s = StringValidator.new(rule)
    s.validated_rule("123").should == :hoge
    s.validated_rule("abc").should == :fuga
    s.validated_rule("xyz").should == nil
  end
end


lib/stringvalidator.rb
# Copyright (C) 2007-2009 TOMITA Masahiro
# mailto:tommy@tmtm.org
#
# = StringValidator
# Author:: TOMITA Masahiro <tommy@tmtm.org>
# License:: Ruby's. see http://www.ruby-lang.org/en/LICENSE.txt
#
# * 文字列が規則に従うかどうかを検査する
# * 規則は Ruby のリテラルで記述する。
#
# == Download
# * http://github.com/tmtm/stringvalidator/downloads
#
# == Install
#  $ make
#  $ make test
#  # make install
#
# == Usage
# StringValidator.validate(_rule_, _str_) は _str_ が _rule_ に適合していれば、_rule_ に適したオブジェクトを返す。
# 適合しなければ StringValidator::Error 例外が発生する。
#
#  StringValidator.validate "abc", "abc"              # => "abc"
#  StringValidator.validate /hoge/, "ahoge"           # => "ahoge"
#  StringValidator.validate Integer, "123"            # => 123
#  StringValidator.validate 123, "123"                # => 123
#  StringValidator.validate 1..255, "128"             # => 128
#  StringValidator.validate 1.0..255, "10.9"          # => 10.9
#  str = "abc"
#  StringValidator.validate ["abc", "def"], str       # => str
#  StringValidator.validate({:length=>3..10}, str)    # => str
#  StringValidator.validate({:minlength=>3, :rule=>/abc/}, str)      # => str
#  StringValidator.validate Proc.new{|a| a == "abc" && 999}, str     # =>999
#  StringValidator.validate Proc.new{|s| Date.parse s}, "2007-10-02" # => Date object
#
# StringValidator.valid?(_rule_, _str_) は _str_ が _rule_ に適合していれば true、そうでなければ false を返す。
#
# 複数ルールの定義
#
#  rules = {
#    :port => 1..65535,
#    :domain => /\A[a-z0-9-]+(\.[a-z0-9-]+)+\z/i,
#  }
#  v = StringValidator.new rules
#  v.valid? :port, "8080"
#  v.valid? :domain, "tmtm.org"
#
# どのルールが適合したか
#
#   rules = {
#     :port => 1..65535,
#     :domain => /\A[a-z0-9]+(\.[a-z0-9]+)+\z/i
#   }
#   v = StringValidator.new rules
#   v.validated_rule "8080"     # => :port
#   v.validated_rule "tmtm.org" # => :domain
#   v.validated_rule "xyz"      # => nil

# == 文字列の正当性チェック
class StringValidator
  # エラーの基底クラス
  class Error < StandardError
    # _value_:: 対象文字列
    # _rule_:: ルールオブジェクト
    def initialize(value=nil, rule=nil)
      @rule = rule
      @value = value
      super self.class.errmsg
    end
    attr_reader :rule, :value

    def self.errmsg()
      @errmsg
    end
  end

  # Integer に適合しない場合
  class Error::NotInteger < Error
    @errmsg = "not integer"
  end
  # Float に適合しない場合
  class Error::NotFloat < Error
    @errmsg = "not float"
  end
  # Range に適合しない場合
  class Error::OutOfRange < Error
    @errmsg = "out of range"
  end
  # Array, Proc, Class に適合しない場合
  class Error::InvalidValue < Error
    @errmsg = "invalid value"
  end
  # Regexp に適合しない場合
  class Error::RegexpMismatch < Error
    @errmsg = "regexp mismatch"
  end
  # Hash(:length, :charlength) に適合しない場合
  class Error::InvalidLength < Error
    @errmsg = "invalid length"
  end
  # Hash(:minlength, :mincharlength) に適合しない場合
  class Error::TooShort < Error
    @errmsg = "too short"
  end
  # Hash(:maxlength, :maxcharlength) に適合しない場合
  class Error::TooLong < Error
    @errmsg = "too long"
  end

  # _rule_:: ルールオブジェクト
  # _str_:: 対象文字列
  #
  # _str_ が _rule_ に適合するか検査する。
  # 適合しない場合、StringValidator::Error 例外が発生する。
  #
  # StringValidator#validate(_rule_, _str_) と同じ。
  #
  def self.validate(rule, str)
    self.new(nil).validate(rule, str)
  end

  # _rule_:: ルールオブジェクト
  # _str_:: 対象文字列
  #
  # validate(_rule_, _str_) が成功すれば true, そうでなければ false を返す。
  #
  def self.valid?(rule, str)
    begin
      self.validate(rule, str)
    rescue Error
      return false
    end
    return true
  end

  # _rule_:: Hash オブジェクト。{:key => rule_object, ...}
  def initialize(rule)
    @rule = rule
  end

  # _rule_:: ルールキー。initialize に与えた Hash のキー
  # _str_:: 対象文字列
  #
  # _str_ が _rule_ に適合するか検査する。
  # 適合しない場合、StringValidator::Error 例外が発生する。
  #
  # === _rule_ の形式
  #
  # ==== Integer (Integer そのもの。Integer オブジェクトではない)
  # Integer(_str_) が成功した場合に _str_ を正当とみなす。
  # Integer オブジェクトを返す。
  # ==== Float (Float そのもの。Float オブジェクトではない)
  # Float(_str_) が成功した場合に _str_ を正当とみなす。
  # Float オブジェクトを返す。
  # ==== Range オブジェクト
  # _rule_ が _str_ を含んでいれば _str_ を正当とみなす。
  # _rule_ に応じて、Integer, Float, String オブジェクトを返す。
  # ==== Regexp オブジェクト
  # _str_ が _rule_ に適合すれば _str_ を正当とみなす。
  # _str_ を返す。
  # ==== Proc オブジェクト
  # _rule_.call(_str_) が例外を発生しなければ _str_ を正当とみなす。
  # _rule_.call(_str_) の結果を返す。
  # ==== Array オブジェクト
  # _rule_ の要素をルールとして評価し、正当な要素が一つでもあれば正当とみなす。
  # 最初に正当になったルールの結果を返す。
  # ==== Hash オブジェクト
  # 複数のルールが指定された場合は、すべてのルールを満たせば正当とみなす。
  # 結果は、:rule(なければ :any, :all)の評価結果。:rule, :any, :all のいずれもなければ str が評価結果となる。
  # [<tt>:any => _array_</tt>]
  #  Array と同じ。
  # [<tt>:all => _array_</tt>]
  #  _array_ の要素をルールとして評価し、すべての要素が正当であれば正当とみなす。
  #  評価結果は、最初の要素のルールの評価結果。
  # [<tt>:rule => _obj_</tt>]
  #  _obj_ をルールとして評価する。
  # [<tt>:length => _integer_ or _range_</tt>]
  #  _str_ の長さ(バイト数)が _integer_ に一致する場合、または _range_ 内であれば正当とみなす。
  # [<tt>:maxlength => _integer_</tt>]
  #  _str_ の長さ(バイト数)が _integer_ 以下であれば正当とみなす。
  # [<tt>:minlength => _integer_</tt>]
  #  _str_ の長さ(バイト数)が _integer_ 以上であれば正当とみなす。
  # [<tt>:charlength => _integer_ or _range_</tt>]
  #  _str_ の長さ(文字数)が _integer_ に一致する場合、または _range_ 内であれば正当とみなす。文字数は $KCODE に依存する。
  # [<tt>:maxcharlength => _integer_</tt>]
  #  _str_ の長さ(文字数)が _integer_ 以下であれば正当とみなす。文字数は $KCODE に依存する。
  # [<tt>:mincharlength => _integer_</tt>]
  #  _str_ の長さ(文字数)が _integer_ 以上であれば正当とみなす。文字数は $KCODE に依存する。
  # ==== Symbol オブジェクト
  # new の引数として渡したハッシュの、キー _rule_ に対応する値をルールとして評価する。
  # ==== Class オブジェクト
  # _rule_.new(_str_) が成功すれば正当とみなす。
  # _rule_.new(_str_) を返す。
  # ==== other
  # _str_ が _rule_.to_s と等しければ正当とみなす。
  # _rule_ を返す。
  #
  def validate(rule, str)
    if rule == Integer then
      begin
        return Integer(str)
      rescue ArgumentError
        raise Error::NotInteger.new(str, rule)
      end
    end
    if rule == Float then
      begin
        return Float(str)
      rescue ArgumentError
        raise Error::NotFloat.new(str, rule)
      end
    end
    case rule
    when Range then
      if rule.first.is_a? Integer and rule.last.is_a? Integer then
        validate(Integer, str)
        raise Error::OutOfRange.new(str, rule) unless rule.include?(str.to_i)
        return str.to_i
      elsif rule.first.is_a? Numeric then
        validate(Float, str)
        raise Error::OutOfRange.new(str, rule) unless rule.include?(str.to_f)
        return str.to_f
      end
      raise  Error::OutOfRange.new(str, rule) unless rule.include?(str)
      return str
    when Regexp then
      raise Error::RegexpMismatch.new(str, rule) unless rule =~ str
      return str
    when Proc then
      begin
        return rule.call(str)
      rescue
        raise Error::InvalidValue.new(str, rule)
      end
      return ret if ret
    when Array then
      rule.each do |i|
        begin
          return self.validate(i, str)
        rescue Error
          nil
        end
      end
      raise Error::InvalidValue.new(str, rule)
    when Hash then
      ret = {}
      rule.each do |k,v|
        case k
        when :any
          ret[k] = self.validate(v, str)
        when :all
          ret[k] = v.map{|i| self.validate(i, str)}.first
        when :rule
          ret[k] = self.validate(v, str)
        when :length
          begin
            self.validate v, str.length.to_s
          rescue Error
            raise Error::InvalidLength.new(str, rule)
          end
        when :maxlength
          raise Error::TooLong.new(str, rule) unless str.length <= v
        when :minlength
          raise Error::TooShort.new(str, rule) unless str.length >= v
        when :charlength
          begin
            self.validate v, str.split(//).length.to_s
          rescue Error
            raise Error::InvalidLength.new(str, rule)
          end
        when :maxcharlength
          raise Error::TooLong.new(str, rule) unless str.split(//).length <= v
        when :mincharlength
          raise Error::TooShort.new(str, rule) unless str.split(//).length >= v
        else
          raise ArgumentError, "Invalid key: #{k}"
        end
      end
      return ret[:rule] || ret[:any] || ret[:all] || str
    when Class then
      begin
        return rule.new(str)
      rescue
        raise Error::InvalidValue.new(str, rule)
      end
    when Symbol then
      raise ArgumentError, "No such rule: #{rule}" unless @rule and @rule.key? rule
      return self.class.validate(@rule[rule], str)
    else
      return rule if rule.to_s == str
    end
    raise Error::InvalidValue.new(str, rule)
  end

  # _rule_:: ルールキー。initialize に与えた Hash のキー
  # _str_:: 対象文字列
  #
  # StringValidator#validate(_rule_, _str) が成功すれば true, そうでなければ false を返す。
  #
  def valid?(rule, str)
    begin
      validate(rule, str)
    rescue Error
      return false
    end
    return true
  end

  # _str_:: 対象文字列
  #
  # initialize に与えた Hash のキーのルールを順に評価し、最初に適合したキー値を返す。
  # 適合したルールがない場合は nil を返す。
  def validated_rule(str)
    @rule.keys.sort{|a,b|a.to_s<=>b.to_s}.each do |k|
      if valid?(@rule[k], str) then
        return k
      end
    end
    return nil
  end

end
